dapps/src/lib.rs

```diff
@@ -101,31 +101,51 @@ impl<F> WebProxyTokens for F where F: Fn(String) -> Option<Origin> + Send + Sync
 }
 
 /// Current supported endpoints.
+#[derive(Default, Clone)]
 pub struct Endpoints {
 	endpoints: endpoint::Endpoints,
+	dapps_path: PathBuf,
+	local_endpoints: Vec<String>,
+	embeddable: Option<ParentFrameSettings>,
 }
 
 impl Endpoints {
 	/// Returns a current list of app endpoints.
 	pub fn list(&self) -> Vec<apps::App> {
-		self.endpoints.iter().filter_map(|(ref k, ref e)| {
+		self.endpoints.read().iter().filter_map(|(ref k, ref e)| {
 			e.info().map(|ref info| apps::App::from_info(k, info))
 		}).collect()
 	}
+	/// Check for any changes in the local dapps folder and update.
+	pub fn refresh_local_dapps(&mut self) {
+		let new_local = apps::fs::local_endpoints(self.dapps_path.clone(), self.embeddable.clone());
+		let (local_endpoints, to_remove) = self.local_endpoints.clone().into_iter().partition(|k| new_local.contains_key(&k.clone()));
```

Comment:
Could be `let (local_endpoints, to_remove) = mem::replace(&mut self.local_endpoints, vec![]).into_iter().parition(...)` to avoid cloning.